### ISSUE TITLE
[a11y]: [Tearsheet]: launcherButtonRef broken with stacked tearsheets

### ISSUE THREAD

**Issue by wkeese** [carbon-design-system/ibm-products#7019]
### Package

@carbon/ibm-products

### Browser

Chrome

### Operating System

MacOS

### Package version

2.59.0

### React version

20.0

### Automated testing tool and ruleset

manual test

### Description

Problem: `launcherButtonRef` doesn't work on stacked tearsheets.

Example markup:

```jsx
      <Tearsheet title="First tearsheet" selectorPrimaryFocus="input" open={isOpen1}>
          <TextInput .../>
          <Button ref={button2Ref} onClick={() => setIsOpen2(true)}>
            Open stacked tearsheet
          </Button>
      </Tearsheet>

      <TearsheetNarrow
        launcherButtonRef={button2Ref}
        actions={[
          {
            key: 1,
            kind: 'primary',
            label: 'Close',
            onClick: () => {
              setIsOpen2(false);
            },
          },
        ]}
        title="Tearsheet 2"
        open={isOpen2}
      >
```

Situation: User opens Tearsheet1, then opens Tearsheet2, then closes Tearsheet2.

Expected result: When Tearsheet 2 is closed, TearsheetShell should focus button2Ref (the button specified by `launcherButtonRef`).

Actual result: When the user closes the stacked tearsheet, TearsheetShell.tsx does focus `launcherButtonRef`.  However, it also focuses the element specified by `selectorPrimaryFocus`, or if `selectorPrimaryFocus` is not specified, then the first element.  Thus a race condition.

Presumably the `selectorPrimaryFocus` related code should only run when a tearsheet is opened.    So I'd expect code like this:

```js
    useEffect(() => {
      if (open) {
        // Focusing the first element or selectorPrimaryFocus element
        claimFocus(firstElement, modalRef, selectorPrimaryFocus);
      }
    }, [open]
```

Instead you have this code that calls `claimFocus()` when Tearsheet 2 is closed, because `depth` has changed:

```js
    useEffect(() => {
      if (open && position === depth) {
        // Focusing the first element or selectorPrimaryFocus element
        claimFocus(firstElement, modalRef, selectorPrimaryFocus);
      }
    }, [
      currentStep,
      depth,
      firstElement,
      modalRef,
      modalRefValue,
      open,
      position,
      selectorPrimaryFocus,
      hasError,
    ]);
```

### Reproduction/example

https://stackblitz.com/edit/github-ykbrpwcg?file=src%2FExample%2FExample.jsx

### Steps to reproduce

1. Open Tearsheet1
2. Open Tearsheet2
3. Close Tearsheet2.
4. Check where focus went.

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/ibm-products/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/ibm-products/issues) for duplicate problems

**Comment by wkeese** [comment]
PS: I suspect this is a regression from 1c918d1605f2a370988d7ade503c1e57e0d43df1.

The old code had this comment which was unfortunately erased:

```
// if we are now the topmost tearsheet, ensure we have focus
```

"Ensure" is the key word.  It means that we should only focus on `selectorPrimaryFocus` if there isn't any other code to focus somewhere better.   In other words,  the nested Tearsheet's `launcherButtonRef` or onClose() handler should take precedence.

Notice how the old code had this condition to achieve that:

```
!modalRefValue.contains(document.activeElement)
```

With that condition, no matter which order the `useEffect()` callbacks run in, the `launcherButtonRef` focus() code from the nested tearsheet won't be overridden by the `selectorPrimaryFocus` code from the original tearsheet.   In the worst case, we might temporarily focus the `selectorPrimaryFocus`, but then immediately switch focus to the nested tearsheet's `launcherButtonRef`.

@makafsal, makes sense?

**Comment by wkeese** [comment]
Note from slack conversation: `launcherButtonRef` is just an example.   If the application defines a (custom) onClose() method that focuses an element, that should take precedence too.

Amazingly on my system I'm something seeing a 350ms delay for claimFocus() to actually run... so there's plenty of time for a race condition to occur.

**Comment by wkeese** [comment]
@makafsal - This should be closed as fixed in 2.64, right?
